statistics/handle/handle_hist.go

```diff
@@ -0,0 +1,359 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handle
+
+import (
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/metrics"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+type StatsLoad struct {
+	sync.Mutex
+	SubCtxs          []sessionctx.Context
+	NeededColumnsCh  chan *NeededColumnTask
+	TimeoutColumnsCh chan *NeededColumnTask
+	workingColMap    map[model.TableColumnID][]chan model.TableColumnID
+}
+
+// NeededColumnTask represents one needed column with expire time.
+type NeededColumnTask struct {
+	TableColumnID model.TableColumnID
+	ToTimeout     time.Time
+	ResultCh      chan model.TableColumnID
+}
+
+// SendLoadRequests send neededColumns requests
+func (h *Handle) SendLoadRequests(sc *stmtctx.StatementContext, neededColumns []model.TableColumnID, timeout time.Duration) {
+	missingColumns := h.genHistMissingColumns(neededColumns)
+	if len(missingColumns) <= 0 {
+		return
+	}
+	sc.StatsLoad.Timeout = timeout
+	sc.StatsLoad.NeededColumns = missingColumns
+	sc.StatsLoad.ResultCh = make(chan model.TableColumnID, len(neededColumns))
+	for _, col := range missingColumns {
+		h.appendNeededColumn(col, sc.StatsLoad.ResultCh, timeout)
+	}
+	sc.StatsLoad.LoadStartTime = time.Now()
+}
+
+// SyncWaitStatsLoad sync waits loading of neededColumns and return false if timeout
+func (h *Handle) SyncWaitStatsLoad(sc *stmtctx.StatementContext) bool {
+	if len(sc.StatsLoad.NeededColumns) <= 0 {
+		return true
+	}
+	defer func() {
+		if sc.StatsLoad.ResultCh != nil {
+			close(sc.StatsLoad.ResultCh)
+		}
+	}()
+	resultCheckMap := map[model.TableColumnID]struct{}{}
+	for _, col := range sc.StatsLoad.NeededColumns {
+		resultCheckMap[col] = struct{}{}
+	}
+	metrics.SyncLoadCounter.Inc()
+	timer := time.NewTimer(sc.StatsLoad.Timeout)
+	defer timer.Stop()
```

Comment:
@tiancaiamao Remove waitgroup and use channel for sync wait as you suggested. Now only one timer is used for one statement for timeout. Should be acceptable, right?